### PR TITLE
feat(cases): record agent case mutations

### PR DESCRIPTION
**Pull request by aaron-tracecat**
## Summary

- Record case mutations when a verified agent-session context is present, normalizing continuation sessions to their root session.
- Cover case creates and updates plus agent-driven comment, tag, task, attachment, linked-row, and dropdown mutations.
- Avoid classifying dropdown values initialized during case creation as a separate update.
- Remove the unused `READ` operation; read-only case access remains unrecorded.
- Treat stale agent-session provenance as optional metadata so it cannot veto the requested case mutation.
- Hold a key-share lock on verified session lineage so concurrent session deletion cannot roll back the case mutation.

## Interaction mapping

| Service operation | Recorded interaction |
| --- | --- |
| Case create | `CREATE` |
| Case update, including no-op and batch updates | `UPDATE` |
| Comment or reply create, edit, or delete | `UPDATE` on the containing case |
| Tag, task, attachment, linked-row, or dropdown mutation | `UPDATE` on the containing case |
| No or stale agent provenance, or read-only operation | None |

Interaction upserts run inside the existing service transaction or per-case batch savepoint. The helper never commits independently, so a failed mutation cannot leave a durable interaction behind.

## Testing

- `uv run pytest tests/unit/test_case_agent_session_interactions.py tests/integration/test_case_agent_session_mutations.py -q` — 10 passed
- Focused case event, task, tag, and attachment service tests — 74 passed
- `uv run pytest tests/unit/test_cases_service.py tests/unit/test_case_comments_service.py -q` — 107 passed
- Ruff, formatting, Basedpyright, Gitleaks, and repository pre-commit hooks passed

## LOC breakdown

| Category | + | - |
| --- | ---: | ---: |
| Logic | 135 | 3 |
| Tests | 377 | 7 |

Linear: [ENG-1710](https://linear.app/tracecat/issue/ENG-1710/record-agent-case-and-comment-mutations-in-case-services)

### CODE REVIEW

File: tracecat/cases/agent_sessions/service.py [hunk 1]
```diff
@@ -18,6 +19,30 @@ class CaseAgentSessionInteractionService(BaseWorkspaceService):
...
+            The inserted or refreshed interaction, or ``None`` when the request
+            has no verified agent-session provenance.
+        """
+        if (agent_session_id := ctx_agent_session_id.get()) is None:
+            return None
+        return await self.record(
+            case_id=case_id,
+            agent_session_id=agent_session_id,
+            operation=operation,
+        )
```
Comment (daryllimyt) [inline tracecat/cases/agent_sessions/service.py]:
```suggestion
        if (agent_session_id := ctx_agent_session_id.get()) is None:
            return None
        try:
            return await self.record(
                case_id=case_id,
                agent_session_id=agent_session_id,
                operation=operation,
            )
        except TracecatNotFoundError:
            logger.warning(
                "Skipping case interaction; agent session not found",
                agent_session_id=agent_session_id,
                case_id=case_id,
            )
            return None
```
Only `TracecatNotFoundError` — a lineage cycle or DB error should still fail the write. Needs `from tracecat.logger import logger` at the top of this module.

**Review by daryllimyt — CHANGES_REQUESTED** [review]
Same-transaction placement looks right — the interaction upsert rides the caller's txn and rolls back with it, and the integration tests cover both directions. One request before merge:

**`record_from_context` should not veto the case/comment write when the agent session row is missing.**

- Auth sets `ctx_agent_session_id` straight from the signed token with no DB check (`tracecat/auth/credentials.py:826`), so a session deleted mid-run leaves a valid token pointing at nothing. Today every `create_case` / `update_case` / comment write from that run then hard-fails with `TracecatNotFoundError` from `resolve_root_session_id`.
- The interaction row is provenance metadata; the mutation is what the caller asked for. Skipping loses nothing recoverable — the FK is `ondelete=CASCADE`, so there is no row to link to anyway.
- `batch_update_cases` already treats this as a per-item soft failure; the single-item paths should be consistent with it.

Suggested change (inline comment below): catch **only** `TracecatNotFoundError` in `record_from_context`, log a warning, return `None`. Keep `TracecatValidationError` (lineage cycle = corrupt data) and DB errors loud.

Please also add one integration test: set the context to a random UUID, assert `create_case` succeeds and zero `CaseAgentSessionInteraction` rows exist.

File: tracecat/cases/agent_sessions/service.py (hunk 1 above)
Comment (aaron-tracecat) [reply tracecat/cases/agent_sessions/service.py]: Fixed — stale session provenance now logs and skips the interaction, with integration coverage.

File: tracecat/cases/service.py [hunk 2]
```diff
@@ -1969,6 +1990,13 @@ class CaseCommentsService(BaseWorkspaceService):
 
     service_name = "case_comments"
 
+    def __init__(self, session: AsyncSession, role: Role | None = None):
+        super().__init__(session, role)
+        self.agent_session_interactions = CaseAgentSessionInteractionService(
+            session=self.session,
+            role=self.role,
+        )
```
[the remark this answers is not in the document]
Comment (aaron-tracecat) [reply tracecat/cases/service.py:1999]: Fixed — child-resource mutations now record at the shared case-event boundary, with task mutations covered directly.

File: tracecat/cases/events.py [hunk 3]
```diff
@@ -10,18 +10,37 @@
...
 from tracecat.service import BaseWorkspaceService
 
 CASE_VIEW_EVENT_DEDUP_WINDOW = timedelta(minutes=5)
+_CHILD_MUTATION_EVENTS = frozenset(
+    {
+        CaseEventType.ATTACHMENT_CREATED,
+        CaseEventType.ATTACHMENT_DELETED,
+        CaseEventType.COMMENT_DELETED,
+        CaseEventType.COMMENT_REPLY_DELETED,
+        CaseEventType.DROPDOWN_VALUE_CHANGED,
```
[the remark this answers is not in the document]
Comment (aaron-tracecat) [reply tracecat/cases/events.py:37]: Fixed — create-time dropdown initialization now suppresses the child `UPDATE` interaction.

File: tracecat/cases/agent_sessions/service.py [hunk 4]
```diff
@@ -18,6 +19,38 @@ class CaseAgentSessionInteractionService(BaseWorkspaceService):
...
+            has no verified agent-session provenance.
+        """
+        if (agent_session_id := ctx_agent_session_id.get()) is None:
+            return None
+        try:
+            return await self.record(
+                case_id=case_id,
+                agent_session_id=agent_session_id,
+                operation=operation,
+            )
```
[the remark this answers is not in the document]
Comment (aaron-tracecat) [reply tracecat/cases/agent_sessions/service.py:45]: Fixed — interaction writes now hold key-share locks on the verified session lineage until commit.

**Review by daryllimyt — APPROVED** [review]
lgtm
